crates/net/dns/src/tree.rs

```diff
@@ -0,0 +1,384 @@
+//! Support for the [EIP-1459 DNS Record Structure](https://eips.ethereum.org/EIPS/eip-1459#dns-record-structure)
+//!
+//! The nodes in a list are encoded as a merkle tree for distribution via the DNS protocol. Entries
+//! of the merkle tree are contained in DNS TXT records. The root of the tree is a TXT record with
+//! the following content:
+//!
+//! ```text
+//! enrtree-root:v1 e=<enr-root> l=<link-root> seq=<sequence-number> sig=<signature>
+//! ```
+//!
+//! where
+//!
+//!    enr-root and link-root refer to the root hashes of subtrees containing nodes and links to
+//! subtrees.
+//!   `sequence-number` is the tree’s update sequence number, a decimal integer.
+//!    `signature` is a 65-byte secp256k1 EC signature over the keccak256 hash of the record
+//! content, excluding the sig= part, encoded as URL-safe base64 (RFC-4648).
+
+use crate::tree::ParseDnsEntryError::{FieldNotFound, UnknownEntry};
+use bytes::Bytes;
+use data_encoding::{BASE32_NOPAD, BASE64URL_NOPAD};
+use enr::{Enr, EnrKey, EnrKeyUnambiguous, EnrPublicKey};
+use reth_primitives::hex;
+use secp256k1::SecretKey;
+use std::{fmt, str::FromStr};
+
+const ROOT_V1_PREFIX: &str = "enrtree-root:v1";
+const LINK_PREFIX: &str = "enrtree://";
+const BRANCH_PREFIX: &str = "enrtree-branch:";
+const ENR_PREFIX: &str = "enr:";
+
+/// Represents all variants
+#[derive(Debug, Clone)]
+#[allow(missing_docs)]
+pub enum DnsEntry<K: EnrKeyUnambiguous> {
+    Root(TreeRootEntry),
+    Link(LinkEntry<K>),
+    Branch(BranchEntry),
+    Node(NodeEntry<K>),
+}
+
+impl<K: EnrKeyUnambiguous> fmt::Display for DnsEntry<K> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            DnsEntry::Root(entry) => entry.fmt(f),
+            DnsEntry::Link(entry) => entry.fmt(f),
+            DnsEntry::Branch(entry) => entry.fmt(f),
+            DnsEntry::Node(entry) => entry.fmt(f),
+        }
+    }
+}
+
+/// Error while parsing a [DnsEntry]
+#[derive(thiserror::Error, Debug)]
+#[allow(missing_docs)]
+pub enum ParseDnsEntryError {
+    #[error("Unknown Entry: {0}")]
+    UnknownEntry(String),
+    #[error("Field {0} not found.")]
+    FieldNotFound(&'static str),
+    #[error("Base64 decoding failed: {0}")]
+    Base64DecodeError(String),
+    #[error("Base32 decoding failed: {0}")]
+    Base32DecodeError(String),
+    #[error("{0}")]
+    RlpDecodeError(String),
+    #[error("{0}")]
+    Other(String),
```

Comment:
Either make these transparent or add context

Reply:
unfortunately the rlp::Error is not reexported from `enr` crate so I'd need to import the crate just for this error variant